tools/harness-automation/autothreadharness/harness_case.py

```diff
@@ -554,28 +603,96 @@ def _test_bed(self):
+                                golden_device_candidates.append(settings.DUT2_DEVICE)
+                                dut_device = settings.DUT2_DEVICE
+                                matched = True
+                                break
+                            for device_item in shield_devices:
+                                if shield_item[0] == mixed_device_item[0] and mixed_device_item[1] == device_item[1]:
+                                    golden_device_candidates.append(device_item)
+                                    shield_devices.remove(device_item)
+                                    matched = True
+                                    break
+                        else:
+                            if 'DUT' in shield_item[0]:
+                                golden_device_candidates.append(settings.DUT_DEVICE)
+                                matched = True
+                                break
+                            for device_item in devices:
+                                if shield_item[0] == mixed_device_item[0] and mixed_device_item[1] == device_item[1]:
+                                    golden_device_candidates.append(device_item)
+                                    devices.remove(device_item)
+                                    matched = True
+                                    break
+                    if not matched:
+                        logger.info('Golden device not enough in : no %s', shield_item)
+                        raise GoldenDeviceNotEnoughError()
+                devices = golden_device_candidates
+                self.add_all_devices = True
             else:
+                for mixed_device_item in topo_mixed_devices:
+                    for device_item in devices:
+                        if mixed_device_item[1] == device_item[1]:
+                            golden_device_candidates.append(device_item)
+                            devices.remove(device_item)
+                            missing_golden_devices.remove(mixed_device_item)
+                            break
+                logger.info('Golden devices in topology config file mapped in settings : %s', golden_device_candidates)
+                if len(topo_mixed_devices) != len(golden_device_candidates):
+                    device_dict = dict()
+                    for missing_device in missing_golden_devices:
+                        if missing_device[1] in device_dict:
+                            device_dict[missing_device[1]] += 1
+                        else:
+                            device_dict[missing_device[1]] = 1
+                    logger.info('Missing Devices: %s', device_dict)
+                    raise GoldenDeviceNotEnoughError()
+                else:
+                    devices = golden_device_candidates
+                    golden_devices_required = len(devices)
+                    logger.info('All case-needed golden devices: %s', json.dumps(devices, indent=2))
+        else:
```

Comment:
How about reducing the duplicate code by taking CV testbed as a special IV testbed from the perspective that the device vendors for specific roles are the same one?

Reply:
If deem the CV testbed as a special IV, need to change the previous Mixed and unique vendor testbed dealing strategy. For each case in CV testbed, will need to read TopologyConfig.txt file once and mapping case role and vendor as well, is it good for time consuming?

Reply:
How long would it take? I am fine to keep it as is if you think it is intolerable.

Reply:
I don't think it is about intolerableness.  I agree with the point that CV can be seen as a special IV. When dealing with testcase on CV testbed, only need to drag the defined number's golden devices from the Golden_Devices list to the testbed list. For IV testbed, have to read the TopologyConfig.txt file and find the testcase needed node roles and their mixed venders first, then mapping the Golden_Devices to choose the right vendor nodes to drag. What I meant is there is not necessary to read the TopologyConfig.txt file for each CV test case which is only needed for IV testbed. What do you think?

Reply:
discussed offline and agreed to enhance these in separate PR.